03.bowling/y-bowling.rb

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+scores = ARGV[0].split(',')
+shots = []
+
+scores.each do |s|
+  shots << (s == 'X' ? 10 : s.to_i)
+  shots << 0 if shots.length < 18 && s == 'X'
+end
+
+frames = shots.each_slice(2).to_a
+
+frames[9] += frames.pop if frames.length == 11
+point = frames.each_with_index.sum do |frame, i|
+  next_frame = frames[i + 1]
+  not_last_frame = i != 9
+  strilke = frame[0] == 10
+  spare = frame.sum == 10
+
+  if not_last_frame && strilke
+    if next_frame.length == 3
+      10 + next_frame[0..1].sum
+    elsif next_frame[0] == 10
+      10 + 10 + frames[i + 2][0]
+    else
+      10 + next_frame.sum
+    end
+  elsif not_last_frame && spare
+    frame.sum + next_frame[0]
+  else
+    frame.sum
+  end
```

Comment:
あくまで参考情報ですが、こんなふうに書くこともできます。

```suggestion
  next frame.sum if i == 9

  next_frame = frames[i + 1]
  strilke = frame[0] == 10
  spare = frame.sum == 10

  if strilke
    if i == 8
      10 + next_frame[0..1].sum
    elsif next_frame[0] == 10
      10 + 10 + frames[i + 2][0]
    else
      10 + next_frame.sum
    end
  elsif spare
    frame.sum + next_frame[0]
  else
    frame.sum
  end
```

ポイントは以下の通りです。

- 最終フレームだけ特別扱いして（`next frame.sum if i == 9`）、21行目、29行目の条件分岐をシンプルにする
- `if next_frame.length == 3` が真になるのは9フレーム目しかないので、`if i == 8` としてしまう（これは好みの問題かも）

Reply:
なるほど、確かに最終フレームを特別扱いするだけで、
そのほかのコードも理解しやすいように書けるようになっているのですね

特別な処理とそうでない処理を切り分けるというのはいろんなところで使えそうなので
使っていこうと思います！